modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/IndexRowComparator.java

```diff
@@ -43,4 +43,12 @@
      * @param idx offset of index key.
      */
     public int compareKey(IndexRow left, IndexRow right, int idx) throws IgniteCheckedException;
+
+    /**
+     * Compare index keys.
+     *
+     * @param left index key.
+     * @param right index key.
+     */
+    public int compareSearchKey(IndexKey left, IndexKey right) throws IgniteCheckedException;
```

Comment:
Question is still open

Reply:
Hi, name of methods are little confusing. I've renamed them. Note, that `IndexRowComparator#compareRow` accepts IndexRow. In some cases we can't replace it with `compareKey` due to additional logic in this method.

Reply:
Why not just call `compareKey` from `compareRow` when both.indexSearchRow() (BTW, is it possible that onlyOne.indexSearchRow()?)

I see we duplicate the following
```
        int ltype = DataType.getTypeFromClass(left.key().getClass());
        int rtype = DataType.getTypeFromClass(right.key().getClass());

        int c = compareValues(wrap(left.key(), ltype), wrap(right.key(), rtype));

        return Integer.signum(c);
```
and wish this can be fixed )

Reply:
> BTW, is it possible that onlyOne.indexSearchRow()

Yes, it's case of finding range in index rows by search row.

Reply:
> I see we duplicate the following and wish this can be fixed

It's duplicated because we put all logic of comparison related to H2 to single place. Actually there is no much of duplicated code there.

Reply:
> Why not just call compareKey from compareRow when both.indexSearchRow()

Because I don't have IndexRow in moment of merging user's criteria. It's just number of index keys of the same field. I should check every boundary of every field. And only after that I can create an IndexRow.